### DISCUSSION TITLE
Cannot get the 'AR is not supported on this device' message to appear. (from the example on https://modelviewer.dev/examples/augmentedreality/#sceneViewer)

### DISCUSSION THREAD

**Discussion by lilzubr** [google/model-viewer#4826] [Q&A] [answered]
Hi, I am new to this so apologies for the very basic question. But I am trying to make a message pop up for when a device isn't compatible with AR so therefore the button doesn't show. I am following the example from [here](https://modelviewer.dev/examples/augmentedreality/#sceneViewer), but am getting an error in the web console:

`(index):35 Uncaught 
TypeError: Cannot read properties of null (reading 'addEventListener')
    at (index):35:7`

This is the code snippet where thats causing the error: 

```javascript
document.querySelector("#model-viewer").addEventListener('ar-status', (event) => {
      if(event.detail.status === 'failed'){
        const error = document.querySelector("#error");
        error.classList.remove('hide');
        error.addEventListener('transitionend',(event) => {
          error.classList.add('hide');
        });
      }
    });
``` 

From looking at model-viewer.js, I believe these events no longer exist, but due to my limited knowledge of javascript, I'm not sure how to proceed and fix it. Any help would be massively appreciated and can provide anymore useful details if needed. Thanks!

**Comment by danilokrulj** [comment]
I think this error appears when you start an AR session on a device where AR is not supported. Did you maybe change model-viewer components id? Since its searching for id "model-viewer".

**Reply by lilzubr** [reply]
Ahhh yes! You were correct! The error has stopped appearing - I must have removed the `id="model-viewer"`

BUT I still can't get the 'AR is not supported on this device' to appear. It seems like the 'hide' element isn't being removed. 

This is my index.html: 
```html
<!DOCTYPE html>
<head>
  <title>Title</title>
  <link rel="icon" href="Icon.jpg" />
</head>
<script
  type="module"
  src="https://unpkg.com/@google/model-viewer/dist/model-viewer.min.js"
></script>
<body>
  <div class="overlay"></div>
  <div class="main">
    <h1>Title</h1>
    <img src="Line.png" alt="line" />
  </div>
  <model-viewer
    id="model-viewer"
    ar
    camera-controls
    touch-action="pan-y"
    src="Model.glb"
    alt="Model"
    camera-target="0m 0.35m 0m"
    camera-orbit="85deg 55deg auto"
    disable-zoom
  >
    <button slot="ar-button">View in AR</button>
    <div id="error" class="hide">AR is not supported on this device</div>
  </model-viewer>
  <script>
    document
      .querySelector("#model-viewer")
      .addEventListener("ar-status", (event) => {
        if (event.detail.status === "failed") {
          const error = document.querySelector("#error");
          error.classList.remove("hide");
          error.addEventListener("transitionend", (event) => {
            error.classList.add("hide");
          });
        }
      });
  </script>
</body>

<style>
  @import url("https://fonts.googleapis.com/css2?family=Libre+Baskerville&display=swap");
  @import url("https://use.typekit.net/atc3jke.css");

  model-viewer {
    width: 100%;
    z-index: 11;
    height: 90vh;
  }

  body {
    overflow: hidden;
    margin: 0;
    touch-action: none;
  }

  .main {
    position: absolute;
    margin-top: 10px;
    text-align: center;
    width: 100%;
    z-index: 12;
  }
  h1 {
    font-family: "Libre Baskerville", serif;
    color: #ffffff;
    font-size: 85px;
    margin-bottom: 20px;
  }
  img {
    width: 80%;
  }

  .overlay {
    position: absolute;
    width: 100%;
    height: 100vh;
    top: 0;
    left: 0;
    background-color: #4d4f59;
    z-index: 10;
    pointer-events: none;
    display: block;
    overflow-y: hidden;
    overflow-x: hidden;
  }

  button {
    position: relative;
    background-color: #db487e;
    border-color: #db487e;
    color: white;
    text-align: center;
    font-family: "proxima-nova", sans-serif;
    font-size: 50px;
    width: 77%;
    padding-top: 45px;
    padding-bottom: 45px;
    margin: 0;
    top: 80%;
    left: 50%;
    -ms-transform: translate(-50%, -50%);
    transform: translate(-50%, -50%);
    border-radius: 12px;
  }

  #error {
    background-color: #ffffffdd;
    border-radius: 16px;
    padding: 16px;
    position: absolute;
    left: 50%;
    top: 50%;
    transform: translate3d(-50%, -50%, 0);
    transition: opacity 0.3s;
    z-index: 15;
  }
  #error.hide {
    opacity: 0;
    visibility: hidden;
    transition: visibility 2s, opacity 1s 1s;
  }

  @media only screen and (orientation: landscape) {
    .main {
      margin-top: 1px;
      text-align: center;
    }
    h1 {
      font-size: 40px;
      margin-bottom: 3px;
    }
    img {
      width: 40%;
    }
    button {
      font-size: 25px;
      width: 38%;
      padding-top: 22px;
      padding-bottom: 22px;
      top: 90%;
    }
  }
</style>

```

**Reply by lilzubr** [reply]
So I can see that `if (event.detail.status === 'failed')` is never hit. So even if you are on a device that doesn't support ar, it doesn't ever seem to be set to 'failed', although the "ar-button" does not appear. I need to try and copy the ar button logic, where it doesn't appear on a non-ar enabled browser. (but make this message appear) - Is that possible?

**Reply by danilokrulj** [reply]
Aaah, but I think the error appears only when the button is hit and the script realizes you actually cant use AR. 
You need something that tells the user before they try entering AR, right?

EDIT: If that is the case maybe you can check if the ar-button is visible, if the device isn't AR compatible the button will be hidden, if the button is hidden you can display a custom message? Its just an idea, not sure if it would work though :)

**Reply by lilzubr** [reply]
Ahh, yes I see. That makes sense! 

Yes thats something I would like to try and do, do you know the best way of checking for that? I'm trying to check for it being hidden but doesn't seem to be hitting that either.

**Reply by danilokrulj** [reply]
I tested like this and i get a log in the console so it should work. (remember to put in the correct id)
```
const arBtn = document.getElementById("ar-button");
if (arBtn.style.visibility = "hidden") {
  do something
}
```

Also I'm not a pro so maybe there is a better way to do this, this was just my first taught :) maybe someone with more knowledge will show up and give a better suggestion

**Reply by lilzubr** [reply]
Amazing, thanks so much!

**Reply by lilzubr** [reply, marked as answer]
Just to update this, I figured out a solid way of getting a AR not supported message to appear every time the AR button does not. 

```javascript
    const arBtn = document.getElementById("arbutton");
    if (arBtn.matches(":not(.enabled)")) {
      const error = document.querySelector("#error");
      error.classList.remove("hide");
    }
    arBtn.addEventListener("click", function () {
      error.classList.add("hide");
    });
``` 
EDIT: need to add an event listener on if the button is clicked, as otherwise when the ar button turns itself off to view in AR, the error will come on.
